lazy_dataset/core.py

```diff
@@ -2313,6 +2397,95 @@ def __iter__(self):
             print(f'Dropped {dropped_count} examples')
 
 
+class CacheDataset(Dataset):
+    def __init__(self, dataset: Dataset, keep_mem_free=None,
+                 immutable_warranty: str = 'pickle') -> None:
+        super().__init__()
+        assert dataset.indexable, (
+            'Cache Dataset only works if dataset is indexable!'
+        )
+        self.dataset = dataset
+        self.cache = {}
+
+        self._keep_mem_free = self._get_memory_size(keep_mem_free)
+        self.immutable_warranty = immutable_warranty
+
+        if immutable_warranty == 'pickle':
+            self._serialize = pickle.dumps
+            self._deserialize = pickle.loads
+        elif immutable_warranty == 'copy':
+            self._serialize = lambda x: x
+            self._deserialize = deepcopy
+        else:
+            raise ValueError(immutable_warranty)
+
+    @property
+    def indexable(self) -> bool:
+        return True
+
+    def keys(self) -> list:
+        return self.dataset.keys()
+
+    @staticmethod
+    def _get_memory_size(keep_mem_free):
+        if keep_mem_free is None:
+            return None
+
+        assert isinstance(keep_mem_free, (list, tuple)) \
+               and len(keep_mem_free) == 2, (
+            f'keep_mem_free has to be a tuple of value and unit, and not '
+            f'{keep_mem_free}'
+        )
+
+        value, unit = keep_mem_free
+
+        import psutil
+
+        if unit == 'GB':
+            return value * 1024 ** 3
+        elif unit == 'fraction':
+            assert 0 <= value < 1, value
+            return psutil.virtual_memory().total * value
+        else:
+            raise ValueError(f'Unknown unit for memory size: {unit}')
+
+    def __getitem__(self, item):
+        if item not in self.cache:
+            # Check if we have enough free memory
+            if self._keep_mem_free is not None:
+                import psutil
+                if psutil.virtual_memory().available <= self._keep_mem_free:
+                    # Return without writing to cache if there is not enough
+                    # free memory
+                    return self.dataset[item]
+
+            self.cache[item] = self._serialize(self.dataset[item])
+        return self._deserialize(self.cache[item])
+
+    def __iter__(self):
+        for k in self.keys():
+            yield self[k]
+
+    def __len__(self):
+        return len(self.dataset)
+
+    def copy(self, freeze: bool = False) -> 'Dataset':
+        if not freeze:
+            import warnings
+            warnings.warn(
+                'Copying a CacheDataset preserves the cache, i.e., the '
+                'already cached part of the dataset will be frozen even if '
+                'freeze=False!'
+            )
+        copy = self.__class__(
+            self.dataset.copy(freeze),
+            keep_mem_free=self._keep_mem_free,
+            immutable_warranty=self.immutable_warranty
+        )
+        copy.cache = self.cache
```

Comment:
Can you add a comment here, that we have to share the cache here.

Reply:
Done